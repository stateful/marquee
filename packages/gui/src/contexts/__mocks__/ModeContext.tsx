import React from 'react';

import modeConfig from './modeConfig.json';

const ModeContext = jest.requireActual('../ModeContext');

export const _setCurrentModeLayout = jest.fn();
export const _removeModeWidget = jest.fn();
export const _duplicateMode = jest.fn();
export const _setModeWidget = jest.fn();
export const _removeMode = jest.fn();
export const _resetModes = jest.fn();
export const _addMode = jest.fn();
export const _setModeName = jest.fn();
export const ModeProvider = ({ children }: any) => (
  <ModeContext.default.Provider value={{
    _setCurrentModeLayout,
    _removeModeWidget,
    _setModeWidget,
    _duplicateMode,
    _removeMode,
    _resetModes,
    _addMode,
    _setModeName,
    modeName: 'default',
    modes: modeConfig,
    thirdPartyWidgets: []
  } as any}>
    {children}
  </ModeContext.default.Provider>
);
export default ModeContext.default;
