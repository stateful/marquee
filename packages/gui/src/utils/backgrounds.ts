/* istanbul ignore next */
export default require.context('../img/backgrounds', true)
