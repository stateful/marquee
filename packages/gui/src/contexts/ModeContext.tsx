import React, {
  createContext,
  useState,
  useEffect,
  useMemo
} from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import { connect, getEventListener, MarqueeEvents } from "@vscode-marquee/utils";
import wrapper, { ThirdPartyWidget } from '@vscode-marquee/widget';
import type { MarqueeWindow, MarqueeInterface, ThirdPartyWidgetOptions } from '@vscode-marquee/utils';
import type { EmojiData } from 'emoji-mart';

import { defaultLayout, defaultEnabledWidgets } from "../constants";
import { Context, Mode, LayoutType, WidgetConfig, State, Configuration, ModeConfig } from "../types";

declare const window: MarqueeWindow;

interface Props {
  children: JSX.Element
}

const ModeContext = createContext<Context>({} as Context);
const WIDGET_ID = '@vscode-marquee/gui';

const ModeProvider = ({ children }: Props) => {
  const eventListener = getEventListener<MarqueeEvents>();
  const [thirdPartyWidgets, setThirdPartyWidgets] = useState([] as WidgetConfig[]);
  const modeState = getEventListener<State & Configuration>(WIDGET_ID);
  const providerValues = connect<State & Configuration>(
    {
      ...window.marqueeStateConfiguration[WIDGET_ID].state,
      ...window.marqueeStateConfiguration[WIDGET_ID].configuration
    },
    modeState
  );

  const _setModeName = (newModeName: string) => {
    if (newModeName !== providerValues.modeName) {
      eventListener.emit('telemetryEvent', {
        eventName: 'switchMode',
        properties: { modeName: newModeName }
      });

      modeState.broadcast({
        modeName: newModeName,
        prevMode: providerValues.modeName,
        modes: providerValues.modes
      });
    }
  };

  const _setCurrentModeLayout = (newLayouts: ReactGridLayout.Layouts) => {
    const newModes: ModeConfig = {
      ...providerValues.modes,
      [providerValues.modeName]: {
        ...providerValues.modes[providerValues.modeName],
        layouts: newLayouts,
      } as Mode,
    };

    if(JSON.stringify(providerValues.modes) === JSON.stringify(newModes)) {
      return;
    }

    providerValues.setModes(newModes);
  };

  const _setModeWidget = (targetMode: string, widgetName: string, value: ReactGridLayout.Layout | Boolean) => {
    providerValues.setModes({
      ...providerValues.modes,
      [targetMode]: {
        ...providerValues.modes[targetMode],
        widgets: {
          ...providerValues.modes[targetMode].widgets,
          [widgetName]: value,
        },
      } as Mode,
    });
  };

  const _removeModeWidget = (widgetName: string) => {
    _setModeWidget(providerValues.modeName || 'default', widgetName, false);
  };

  const _resetModes = () => {
    providerValues.setModes(undefined as any);
    providerValues.setModeName('default');
    providerValues.setPrevMode(null);
  };

  const _removeMode = (removeModeName: string) => {
    //it must exist
    if (!providerValues.modes[removeModeName]) {
      return;
    }

    //if we are deleting the selected mode
    if (removeModeName === providerValues.modeName) {
      providerValues.setModeName("default");
    }

    providerValues.setModes(Object.assign(
      {},
      ...Object.entries(providerValues.modes)
        .filter(([k]) => k !== removeModeName)
        .map(([k, v]) => ({ [k]: v }))
    ));
  };

  const _addModeWithParams = (newModeName: string, emoji: EmojiData, layouts: LayoutType, widgets: Record<string, boolean>) => {
    newModeName = newModeName.toLowerCase();

    if (!providerValues.modes[newModeName]) {
      let newModeObj: Mode = {
        layouts: layouts,
        widgets: widgets,
        icon: emoji,
      };
      providerValues.setModes({
        ...providerValues.modes,
        [newModeName]: newModeObj,
      });
    }
  };

  const _duplicateMode = (oldModeName: string, newModeName: string, newEmoji: EmojiData) => {
    let prevLayouts = Object.assign({}, providerValues.modes[oldModeName].layouts);
    let prevWidgets = Object.assign({}, providerValues.modes[oldModeName].widgets);
    _addModeWithParams(newModeName, newEmoji, prevLayouts, prevWidgets);
  };

  const _addMode = (newModeName: string, emoji: EmojiData) => {
    _addModeWithParams(
      newModeName,
      emoji,
      defaultLayout,
      defaultEnabledWidgets
    );
  };

  window.marqueeExtension = useMemo(() => ({
    defineWidget: (
      widgetOptions: ThirdPartyWidgetOptions,
      constructor: CustomElementConstructor,
      options?: ElementDefinitionOptions
    ) => {
      customElements.define(widgetOptions.name, constructor, options);
      setThirdPartyWidgets([...thirdPartyWidgets, {
        name: widgetOptions.name,
        icon: <FontAwesomeIcon icon={widgetOptions.icon} />,
        label: widgetOptions.label,
        tags: widgetOptions.tags,
        description: widgetOptions.description,
        component: wrapper(ThirdPartyWidget, widgetOptions.name),
      }]);
    }
  } as MarqueeInterface), [thirdPartyWidgets]);

  useEffect(() => {
    eventListener.emit(
      'updateWidgetDisplay',
      providerValues.modes[providerValues.modeName].widgets
    );

    return () => { eventListener.removeAllListeners(); };
  }, []);

  return (
    <ModeContext.Provider
      value={{
        ...providerValues,
        thirdPartyWidgets,
        setThirdPartyWidgets,

        _setModeName: _setModeName,
        _setCurrentModeLayout: _setCurrentModeLayout,
        _setModeWidget: _setModeWidget,
        _addMode: _addMode,
        _removeMode: _removeMode,
        _resetModes: _resetModes,
        _duplicateMode: _duplicateMode,
        _removeModeWidget: _removeModeWidget,
      }}
    >
      {children}
    </ModeContext.Provider>
  );
};

export default ModeContext;
export { ModeProvider };
