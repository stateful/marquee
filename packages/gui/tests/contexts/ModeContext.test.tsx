import React, { useContext } from 'react'
import { act } from 'react-dom/test-utils'
import { render } from '@testing-library/react'
// @ts-expect-error
import { MarqueeWindow, getEventListener, providerValues } from '@vscode-marquee/utils'
import { faBrain } from '@fortawesome/free-solid-svg-icons/faBrain'

import ModeContext, { ModeProvider } from '../../src/contexts/ModeContext'

declare const window: MarqueeWindow

let _thirdPartyWidgets: any
let _modes: any
let __addMode: any
let __removeMode: any
let __resetModes: any
let __setModeName: any

test('WidgetLayout filters non display widgets', () => {
  expect(window.marqueeExtension).not.toBeTruthy()
  const TestComponent = () => {
    const { thirdPartyWidgets, modes, _resetModes, _removeMode, _addMode, _setModeName } = useContext(ModeContext)
    _thirdPartyWidgets = thirdPartyWidgets
    _modes = modes
    __addMode = _addMode
    __removeMode = _removeMode
    __resetModes = _resetModes
    __setModeName = _setModeName
    return (<div>Hello World</div>)
  }

  render(<ModeProvider><TestComponent /></ModeProvider>)
  expect(window.marqueeExtension).toBeTruthy()
  expect(typeof window.marqueeExtension.defineWidget).toBe('function')

  expect(_thirdPartyWidgets).toHaveLength(0)
  act(() => {
    window.marqueeExtension.defineWidget({
      name: 'foo-bar',
      icon: faBrain,
      label: 'some label',
      tags: ['some', 'tags'],
      description: 'some description'
    }, class Foobar extends HTMLElement {})
  })
  expect(_thirdPartyWidgets).toHaveLength(1)

  act(() => { __setModeName('work') })
  const listener = getEventListener()
  expect(listener.broadcast).toBeCalledWith({
    modeName: 'work',
    modes: expect.any(Object)
  })

  act(() => { __addMode('foobar', null) })
  expect(_modes).toMatchSnapshot()

  act(() => { __removeMode('foobar') })
  expect(providerValues.setModes).toBeCalledTimes(1)

  listener.emit = jest.fn()
  act(() => { __resetModes() })
  expect(listener.emit).toBeCalledWith('clear', undefined)
})
