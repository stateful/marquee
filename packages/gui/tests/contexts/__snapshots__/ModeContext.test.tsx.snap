// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetLayout filters non display widgets 1`] = `
Object {
  "default": Object {
    "layouts": Object {
      "lg": Array [
        Object {
          "h": 12,
          "i": "welcome",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 8,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 8,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 4,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 12,
        },
        Object {
          "h": 13,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 24,
        },
        Object {
          "h": 13,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 6,
          "y": 24,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 4,
          "y": 0,
        },
      ],
      "md": Array [
        Object {
          "h": 12,
          "i": "welcome",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 6,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 7,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 12,
        },
        Object {
          "h": 13,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 0,
          "y": 24,
        },
        Object {
          "h": 13,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 5,
          "y": 24,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 3,
          "y": 0,
        },
      ],
      "sm": Array [
        Object {
          "h": 12,
          "i": "welcome",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 24,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 24,
        },
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 12,
        },
        Object {
          "h": 13,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 36,
        },
        Object {
          "h": 13,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 49,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 0,
        },
      ],
      "xs": Array [
        Object {
          "h": 11,
          "i": "welcome",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 2,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 23,
        },
        Object {
          "h": 11,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 2,
          "x": 2,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 74,
        },
        Object {
          "h": 13,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 61,
        },
        Object {
          "h": 13,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 35,
        },
        Object {
          "h": 13,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 48,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 11,
        },
      ],
      "xxs": Array [
        Object {
          "h": 11,
          "i": "welcome",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 35,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 11,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 86,
        },
        Object {
          "h": 13,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 73,
        },
        Object {
          "h": 13,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 47,
        },
        Object {
          "h": 13,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 60,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 23,
        },
      ],
    },
    "widgets": Object {
      "github": true,
      "news": true,
      "notes": true,
      "projects": true,
      "snippets": true,
      "todo": true,
      "weather": true,
      "welcome": true,
    },
  },
  "play": Object {
    "icon": Object {
      "colons": ":beach_with_umbrella:",
      "id": "beach_with_umbrella",
      "name": "Beach with Umbrella",
      "native": "üèñÔ∏è",
      "short_names": Array [
        "beach_with_umbrella",
      ],
      "unified": "1f3d6-fe0f",
    },
    "layouts": Object {
      "lg": Array [
        Object {
          "h": 14,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 6,
          "y": 15,
        },
        Object {
          "h": 14,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 15,
        },
        Object {
          "h": 14,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 4,
          "y": 29,
        },
        Object {
          "h": 15,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 0,
        },
        Object {
          "h": 15,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 6,
          "y": 0,
        },
        Object {
          "h": 15,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 29,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 41,
        },
      ],
      "md": Array [
        Object {
          "h": 14,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 5,
          "y": 12,
        },
        Object {
          "h": 14,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 0,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 37,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 5,
          "y": 0,
        },
        Object {
          "h": 11,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 10,
          "x": 0,
          "y": 26,
        },
        Object {
          "h": 12,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 49,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 61,
        },
      ],
      "sm": Array [
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 11,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 11,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 36,
        },
        Object {
          "h": 11,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 0,
        },
        Object {
          "h": 13,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 23,
        },
        Object {
          "h": 11,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 48,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 60,
        },
      ],
      "xs": Array [
        Object {
          "h": 13,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 37,
        },
        Object {
          "h": 13,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 24,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 62,
        },
        Object {
          "h": 11,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 13,
        },
        Object {
          "h": 12,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 50,
        },
        Object {
          "h": 13,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 74,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 86,
        },
      ],
      "xxs": Array [
        Object {
          "h": 13,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 37,
        },
        Object {
          "h": 13,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 24,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 62,
        },
        Object {
          "h": 11,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 13,
        },
        Object {
          "h": 12,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 50,
        },
        Object {
          "h": 13,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 74,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 86,
        },
      ],
    },
    "widgets": Object {
      "github": true,
      "news": true,
      "notes": false,
      "projects": false,
      "snippets": false,
      "todo": false,
      "weather": true,
      "welcome": true,
    },
  },
  "work": Object {
    "icon": Object {
      "colons": ":briefcase:",
      "id": "briefcase",
      "name": "Briefcase",
      "native": "üíº",
      "short_names": Array [
        "briefcase",
      ],
      "unified": "1f4bc",
    },
    "layouts": Object {
      "lg": Array [
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 28,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 40,
        },
        Object {
          "h": 13,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 8,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 52,
        },
        Object {
          "h": 15,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 6,
          "y": 13,
        },
        Object {
          "h": 13,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 15,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 13,
        },
        Object {
          "h": 13,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 4,
          "y": 0,
        },
      ],
      "md": Array [
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 26,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 38,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 50,
        },
        Object {
          "h": 14,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 5,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 14,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 5,
          "x": 0,
          "y": 12,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 6,
          "y": 0,
        },
      ],
      "sm": Array [
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 47,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 59,
        },
        Object {
          "h": 11,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 3,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 71,
        },
        Object {
          "h": 11,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 36,
        },
        Object {
          "h": 11,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 3,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 13,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 23,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 6,
          "x": 0,
          "y": 11,
        },
      ],
      "xs": Array [
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 60,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 72,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 25,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 84,
        },
        Object {
          "h": 11,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 49,
        },
        Object {
          "h": 13,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 37,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 13,
        },
      ],
      "xxs": Array [
        Object {
          "h": 12,
          "i": "news",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 60,
        },
        Object {
          "h": 12,
          "i": "github",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 72,
        },
        Object {
          "h": 12,
          "i": "todo",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 25,
        },
        Object {
          "h": 12,
          "i": "weather",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 84,
        },
        Object {
          "h": 11,
          "i": "notes",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 49,
        },
        Object {
          "h": 13,
          "i": "welcome",
          "minH": 9,
          "minW": 3,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 12,
          "i": "snippets",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 37,
        },
        Object {
          "h": 12,
          "i": "projects",
          "minH": 6,
          "minW": 2,
          "moved": false,
          "static": false,
          "w": 4,
          "x": 0,
          "y": 13,
        },
      ],
    },
    "widgets": Object {
      "github": false,
      "news": false,
      "notes": true,
      "projects": true,
      "snippets": true,
      "todo": true,
      "weather": false,
      "welcome": true,
    },
  },
}
`;
